plutus-tx-plugin/src/PlutusTx/Options.hs

```diff
@@ -0,0 +1,234 @@
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes       #-}
+{-# LANGUAGE RankNTypes        #-}
+{-# LANGUAGE StrictData        #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TypeApplications  #-}
+
+module PlutusTx.Options where
+
+import PlutusIR.Compiler qualified as PIR
+import PlutusTx.Compiler.Types
+
+import Control.Exception
+import Control.Lens
+import Data.Bifunctor (first)
+import Data.Either.Validation
+import Data.Foldable (foldl', toList)
+import Data.List qualified as List
+import Data.List.NonEmpty (NonEmpty)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Proxy
+import Data.Text (Text)
+import Data.Text qualified as Text
+import GhcPlugins qualified as GHC
+import PyF (fmt)
+import Text.Read (readMaybe)
+import Type.Reflection
+
+data PluginOptions = PluginOptions
+    { _poDoTypecheck                    :: Bool
+    , _poDeferErrors                    :: Bool
+    , _poContextLevel                   :: Int
+    , _poDumpPir                        :: Bool
+    , _poDumpPlc                        :: Bool
+    , _poDumpUPlc                       :: Bool
+    , _poOptimize                       :: Bool
+    , _poPedantic                       :: Bool
+    , _poVerbose                        :: Bool
+    , _poDebug                          :: Bool
+    , _poMaxSimplifierIterations        :: Int
+    , _poDoSimplifierUnwrapCancel       :: Bool
+    , _poDoSimplifierBeta               :: Bool
+    , _poDoSimplifierInline             :: Bool
+    , _poDoSimplifierRemoveDeadBindings :: Bool
+    , _poProfile                        :: ProfileOpts
+    , _poCoverageAll                    :: Bool
+    , _poCoverageLocation               :: Bool
+    , _poCoverageBoolean                :: Bool
+    , -- Setting to `True` defines `trace` as `\_ a -> a` instead of the builtin version.
+      -- Which effectively ignores the trace text.
+      _poRemoveTrace                    :: Bool
+    }
+
+makeLenses ''PluginOptions
+
+type OptionKey = Text
+type OptionValue = Text
+
+-- | A plugin option definition for a `PluginOptions` field of type @a@.
+data PluginOption
+    = forall a.
+        PluginOption
+        (TypeRep a)
+        -- ^ `TypeRep` used for pretty printing the option.
+        (Maybe OptionValue -> Validation ParseError (a -> a))
+        -- ^ Consumes an optional value, and either updates the field or reports an error.
+        (Lens' PluginOptions a)
+        -- ^ Lens focusing on the field. This is for modifying the field, as well as
```

Comment:
>  perhaps we should be parsing our options into a simple record with a field for each option

Yeah I think this is the right thing to do, just because it's more friendly to automatically generating docs.

> No need for the parser to produce an update function

Update functions can be useful (at least to the other plugin I worked on where values from multiple options with the same key are concatenated into lists). Another example is to increase the verbosity for each `verbose`